tests/cpp/test_inlining.cpp

```diff
@@ -0,0 +1,325 @@
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> shape1({3, 4, 5});
+  std::vector<int64_t> shape2({3, 4 * 5});
+
+  // Inner normalization pattern
+
+  // [i0, i1, i2]
+  auto tv0 = makeConcreteTensor(shape1);
+  fusion.addInput(tv0);
+
+  // [i0, i1*i2]
+  auto tv1 = reshape(tv0, shape1, shape2);
+  // [i0, r1*r2]
+  auto tv2 = sum(tv1, {1});
+  // [i0]
+  auto tv3 = add(tv2, fusion.oneVal());
+  // [i0, b3]
+  auto tv4 = broadcast(tv3, {false, true});
+  // [i0, i1*i2]
+  auto tv5 = sub(tv1, tv4);
+  // [i0, i1, i2]
+  auto tv6 = reshape(tv5, shape2, shape1);
+  fusion.addOutput(tv6);
+
+  MaxPosCalculator calc;
+
+  auto isAllowedID = [&calc](TensorView* tv, int64_t axis) -> bool {
+    return calc.isAllowedID(
+        tv->getLoopDomain().at(axis),
+        tv,
+        /*best_effort=*/true,
+        /*allow_reduction=*/false,
+        /*allow_vectorize=*/false,
+        /*allow_unmappable=*/false);
+  };
+
+  // First, check isAllowedID without manipulating loop domains. tv1
+  // is a persistent tensor. The inner domain needs to be persistent.
+  EXPECT_TRUE(isAllowedID(tv1, 0)) << tv1->getLoopDomain().at(0)->toString();
+  // The inner domain should not be allowed since it's a persistent domain
+  EXPECT_FALSE(isAllowedID(tv1, 1)) << tv1->getLoopDomain().at(1)->toString();
+
+  // Set [i0, i1, i2] as the loop domain of each of tensors. i1 and i2
+  // become pesistent domains.
+
+  // tv1
+  {
+    tv1->setLoopDomain(tv1->getRootDomain());
+    EXPECT_TRUE(isAllowedID(tv1, 0)) << tv1->getLoopDomain().at(0)->toString();
+    // Persistent domain
+    EXPECT_FALSE(isAllowedID(tv1, 1)) << tv1->getLoopDomain().at(1)->toString();
+    // Persistent domain
+    EXPECT_FALSE(isAllowedID(tv1, 2)) << tv1->getLoopDomain().at(2)->toString();
+  }
+
+  // tv2
+  {
+    std::vector<IterDomain*> loop_domain{
+        tv2->getLogicalDomain().at(0),
+        tv1->getLoopDomain().at(1)->cloneWithoutRFactor(),
+        tv1->getLoopDomain().at(2)->cloneWithoutRFactor()};
+    IrBuilder::create<Merge>(
+        tv2->getLogicalDomain().at(1), loop_domain[1], loop_domain[2]);
+    tv2->setLoopDomain(loop_domain);
+
+    EXPECT_TRUE(isAllowedID(tv2, 0)) << tv2->getLoopDomain().at(0)->toString();
+    // Reduction domain
+    EXPECT_FALSE(isAllowedID(tv2, 1)) << tv2->getLoopDomain().at(1)->toString();
+    // Reduction domain
+    EXPECT_FALSE(isAllowedID(tv2, 2)) << tv2->getLoopDomain().at(2)->toString();
+  }
+
+  // tv3
+  {
+    std::vector<IterDomain*> loop_domain{
+        tv1->getLoopDomain().at(0)->cloneWithoutRFactor(),
+        tv1->getLoopDomain().at(1)->cloneWithoutRFactor(),
+        tv1->getLoopDomain().at(2)->cloneWithoutRFactor()};
+    IrBuilder::create<Merge>(
+        tv3->getLogicalDomain().at(0), loop_domain[1], loop_domain[2]);
+    tv3->setLoopDomain(loop_domain);
+    for (const auto i : c10::irange(3)) {
+      EXPECT_TRUE(isAllowedID(tv3, i))
+          << tv3->getLoopDomain().at(i)->toString();
+    }
```

Comment:
If I have
```
producer: [I0, I1]
```
and
```
consumer: [I0*I1]
```
Then can we inline `producer` at 1?

Reply:
No, the producer also needs the merge of `I0` and `I1`.

Note that `isAllowedID` doesn't consider producer-consumer relationships. It's only about just the given ID itself, except for the persistence constraint by `ComputeAtLogicalDomainMap`.